processing/src/main/java/io/druid/segment/DimensionQueryHelper.java

```diff
@@ -0,0 +1,359 @@
+   */
+  public ValueMatcher getValueMatcher(ColumnSelectorFactory cursor, Comparable value);
+
+
+  /**
+   * Create a predicate-based ValueMatcher, used for filtering by QueryableIndexStorageAdapter and FilteredAggregatorFactory.
+   *
+   * @param cursor ColumnSelectorFactory for creating dimension value selectors
+   * @param predicateFactory A DruidPredicateFactory that provides the filter predicates to be matched
+   * @return A ValueMatcher that applies the predicate for this DimensionQueryHelper's value type from the predicateFactory
+   */
+  public ValueMatcher getValueMatcher(ColumnSelectorFactory cursor, final DruidPredicateFactory predicateFactory);
+
+
+  /**
+   * Used by CardinalityAggregator.
+   *
+   * Retrieve the current row from dimSelector and add the row values to the hasher.
+   *
+   * @param dimSelector Dimension value selector
+   * @param hasher Hasher used for cardinality aggregator calculations
+   */
+  public void hashRow(Object dimSelector, Hasher hasher);
+
+
+  /**
+   * Used by CardinalityAggregator.
+   *
+   * Retrieve the current row from dimSelector and add the row values to the hasher.
+   * @param dimSelector Dimension value selector
+   * @param collector HLL collector used for cardinality aggregator calculations
+   */
+  public void hashValues(Object dimSelector, HyperLogLogCollector collector);
+
+
+  /**
+   * Used by GroupByEngine.
+   *
+   * Return the size, in bytes, of this dimension's values in the grouping key.
+   *
+   * For example, a String implementation would return 4, the size of an int.
+   *
+   * @return size, in bytes, of this dimension's values in the grouping key.
+   */
+  public int getGroupingKeySize();
+
+
+  /**
+   * Used by GroupByEngine.
+   *
+   * A grouping key contains a concatenation of byte[] representations of dimension values.
+   *
+   * When comparing two grouping keys, the individual dimension values will be compared with comparators
+   * provided by the query helper.
+   *
+   * @return A comparator suitable for comparing byte representations of this dimension's type of values.
+   */
+  public Comparator<byte[]> getGroupingKeyByteComparator();
+
+
+  /**
+   * Used by GroupByEngine.
+   *
+   * Perform a relative read on a grouping key ByteBuffer to retrieve a single dimension value, and
+   * add the retrieved value to a GroupBy result map.
+   *
+   * An implementation may choose to not add anything to the result map
+   * (e.g., as the String implementation does for empty rows)
+   *
+   * @param theEvent Result map for the GroupBy query being served
+   * @param outputName The output name of this dimension for the GroupBy query being served, as specified in the DimensionSpec
+   * @param dimSelector Dimension value selector, used for value lookups if needed
+   * @param keyBuffer Grouping key, already positioned at this dimension's offset
+   */
+  public void readDimValueFromGroupingKey(
+      Map<String, Object> theEvent,
```

Comment:
I think there is a general convention in Druid to place "output" parameters after "input" parameters? `theEvent` is an "output" parameter here, so maybe it should be the last parameter?